firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -159,66 +159,91 @@ public static FirebaseInstallations getInstance(@NonNull FirebaseApp app) {
-    return persistedFidEntry;
+    return taskCompletionSource.getTask();
   }
 
-  private static boolean persistedFidMissingOrInErrorState(PersistedFidEntry persistedFidEntry) {
-    return persistedFidEntry == null
-        || persistedFidEntry.getRegistrationStatus() == RegistrationStatus.REGISTER_ERROR;
+  private Task<InstallationTokenResult> addGetAuthTokenListener(int authTokenOption) {
+    TaskCompletionSource<InstallationTokenResult> taskCompletionSource =
+        new TaskCompletionSource<>();
+    String oldAuthToken =
+        persistedFid.readPersistedFidEntryValue().getAuthToken() == null
+            ? ""
+            : persistedFid.readPersistedFidEntryValue().getAuthToken();
+    StateListener l =
+        new GetAuthTokenListener(utils, taskCompletionSource, oldAuthToken, authTokenOption);
+    synchronized (lock) {
+      listeners.add(l);
+    }
+    return taskCompletionSource.getTask();
   }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> orElse(@NonNull Supplier<T> supplier) {
-    return t -> {
-      if (t.isSuccessful()) {
-        return (T) t.getResult();
+  private void triggerOnStateReached(PersistedFidEntry persistedFidEntry) {
+    synchronized (lock) {
+      Iterator<StateListener> it = listeners.iterator();
+      while (it.hasNext()) {
+        StateListener l = it.next();
+        boolean doneListening = l.onStateReached(persistedFidEntry);
+        if (doneListening) {
+          it.remove();
+        }
       }
-      return supplier.get();
-    };
+    }
   }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> awaitFidRegistration(
-      @NonNull Supplier<T> supplier, AwaitListener listener) {
-    return t -> {
-      // Waiting for Task that registers FID on the FIS Servers
-      listener.await(AWAIT_TIMEOUT_IN_SECS, TimeUnit.SECONDS);
-      return supplier.get();
+  private final Runnable doRegistration(int authTokenOption) {
+    return () -> {
+      try {
+        PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+
+        // New FID needs to be created
+        if (persistedFidEntry.isErrored() || persistedFidEntry.isNotGenerated()) {
+          String fid = utils.createRandomFid();
+          persistFid(fid);
+          persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+        }
+
+        triggerOnStateReached(persistedFidEntry);
+
+        // FID needs to be registered
+        if (persistedFidEntry.isUnregistered()) {
+          registerAndSaveFid(persistedFidEntry);
+          persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+        }
+
+        // Don't notify the listeners at this point; we might as well make ure the auth token is up
+        // to date before letting them know.
+
+        // Refresh Auth token if:
+        // 1. Invalid auth token
+        // 2. If FORCE_REFRESH with a pending listener
+        synchronized (lock) {
+          if ((authTokenOption == FORCE_REFRESH && !listeners.isEmpty())
```

Comment:
I get why we have this `!listeners.isEmpty()` check, but it feels like it's a symptom of what we're trying to check (is anyone still waiting for a state change?) instead of the actual thing we want to check (whether our current auth token needs to be refreshed)

Reply:
Makes sense. Updated. PTAL. Thanks